birdhouse/components/weaver/config/magpie/weaver_hooks.py

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+These hooks will be running within Twitcher, using MagpieAdapter context, applied for Weaver requests.
+
+The code below can make use of any package that is installed by Magpie/Twitcher.
+"""
+
+import json
+from typing import TYPE_CHECKING
+
+from magpie.api.management.resource import resource_utils as ru
+from magpie.api.requests import get_user
+from magpie.constants import get_constant
+from magpie.permissions import Access, Permission
+from magpie.utils import get_header
+
+if TYPE_CHECKING:
+    from pyramid.request import Request
+    from pyramid.response import Response
+
+    from magpie.adapter import HookContext
+
+
+def is_admin(request):
+    # type: (Request) -> bool
+    admin_group = get_constant("MAGPIE_ADMIN_GROUP", settings_container=request)
+    if not request.user:  # no user authenticated (public)
+        return False
+    return admin_group in [group.group_name for group in request.user.groups]
+
+
+def add_x_wps_output_context(request):
+    # type: (Request) -> Request
+    """
+    Apply the ``X-WPS-Output-Context`` for saving outputs in the user-context WPS-outputs directory.
+    """
+    header = get_header("X-WPS-Output-Context", request.headers)
+    # if explicitly provided, ensure it is permitted (admin allow any, otherwise self-user reference only)
+    if header is not None:
+        if request.user is None:
+            header = "public"
+        else:
+            if not is_admin(request):
+                # override disallowed writing to other location
+                # otherwise, up to admin to have writen something sensible
+                header = "user-" + str(request.user.id)
```

Comment:
Just curious, if it is indeed an admin user, what other values than `"user-" + str(request.user.id)` can `X-WPS-Output-Context` have?

Reply:
It can be any valid directory name. 
Directories will be formed (and created by Weaver if missing) using this format:

`<wps_output_dir>/<wps_output_context>/<job_id>/<output-files>`

If `X-WPS-Output-Context` is not present instead (current behavior), it is:

`<wps_output_dir>/<job_id>/<output-files>`

So basically, the hook enforces outputs to:
- `<wps_output_dir>/user-<id>/<job_id>/<output-files>` for a non-admin
- `<wps_output_dir>/public/<job_id>/<output-files>` for no-auth (process will have to be public as well to allow execute)
- `<wps_output_dir>/**anywhere**/<job_id>/<output-files>` for an admin (possibly using literal `user-<id>` if desired)

Reply:
Oh, I did not realized `X-WPS-Output-Context` will be used part of the wps output dir, thanks.